### ISSUE TITLE
Match statement fails to type narrow Generic T while if isinstance works (strict mode)

### ISSUE THREAD

**Issue by GDYendell** [microsoft/pyright#9089]
**Describe the bug**
I have a case where match statement does not type narrow the generic type of instances, while an equivalent if statement with `isinstance` works correctly.

This is the output of pyright on the code snippet below.

```
/workspaces/FastCS/my_typing.py
  /workspaces/FastCS/my_typing.py:24:49 - error: Function with declared return type "T@get_value_match" must return value on all code paths
    Type "None" is not assignable to type "T@get_value_match" (reportReturnType)
  /workspaces/FastCS/my_typing.py:27:25 - information: Type of "ab" is "Attribute[T@get_value_match]"
  /workspaces/FastCS/my_typing.py:28:25 - information: Type of "vb" is "bool*"
  /workspaces/FastCS/my_typing.py:31:25 - information: Type of "ai" is "Attribute[T@get_value_match]"
  /workspaces/FastCS/my_typing.py:32:25 - information: Type of "vi" is "bool* | int*"
  /workspaces/FastCS/my_typing.py:35:25 - information: Type of "af" is "Attribute[T@get_value_match]"
  /workspaces/FastCS/my_typing.py:36:25 - information: Type of "vf" is "float*"
  /workspaces/FastCS/my_typing.py:39:25 - information: Type of "other" is "Attribute[T@get_value_match]"
  /workspaces/FastCS/my_typing.py:45:21 - information: Type of "vb" is "bool*"
  /workspaces/FastCS/my_typing.py:49:21 - information: Type of "vi" is "int*"
  /workspaces/FastCS/my_typing.py:53:21 - information: Type of "vf" is "float*"
1 error, 0 warnings, 10 informations 
```

It knows the `vb` / `vi` / `vf` values are `bool` / `int` / `float` and therefore knows that `T` can be those types (because it doesn't give a `Pattern will never match ...` error), but it doesn't narrow `T` when revealing the type of the attribute and (presumably for the same reason) doesn't know that all cases are covered.

If I am doing something wrong here, any advice would be much appreciated! I would also be interested to know exactly what the `*` syntax on the revealed types means.

**Code or Screenshots**

I think this is a minimal reduced version of the original code, which has made all but one of the pyright errors in the original code disappear. It may be that they all stemmed from this one, or there may be other problems. I can provide slightly less minimal versions if it would be useful. For example, `Attribute` doesn't actually take a `value` `T`, it takes a `datatype` `DataType[T]` and then there are `Bool(DataType[bool])`, `Int(DataType[int])` classes, which are then matched using e.g. `case Attribute(datatype=Bool())`. This was giving `Pattern will never match ...` errors. In the original code I can't just `isinstance` the individual field, I need to narrow the whole `Attribute` instance.

```python
from typing import Generic, TypeVar, reveal_type

T = TypeVar("T", bool, float, int)


class Attribute(Generic[T]):
    def __init__(self, value: T) -> None:
        self.value = value


def create_attribute(value: T) -> Attribute[T]:
    match value:
        case bool() as bool_value:
            ab = Attribute(bool_value)
            return ab
        case int() as int_value:
            ai = Attribute(int_value)
            return ai
        case float() as float_value:
            af = Attribute(float_value)
            return af


def get_value_match(attribute: Attribute[T]) -> T:
    match attribute:
        case Attribute(value=bool() as vb) as ab:
            reveal_type(ab)
            reveal_type(vb)
            return vb
        case Attribute(value=int() as vi) as ai:
            reveal_type(ai)
            reveal_type(vi)
            return vi
        case Attribute(value=float() as vf) as af:
            reveal_type(af)
            reveal_type(vf)
            return vf
        case other:
            reveal_type(other)


def get_value_if(attribute: Attribute[T]) -> T:
    if isinstance(attribute.value, bool):
        vb: bool = attribute.value
        reveal_type(vb)
        return vb
    elif isinstance(attribute.value, int):
        vi: int = attribute.value
        reveal_type(vi)
        return vi
    else:  # if isinstance(attribute.value, float): not needed as pyright knows it is exhaustive
        vf: float = attribute.value
        reveal_type(vf)
        return vf


ab = create_attribute(True)
ai = create_attribute(1)
af = create_attribute(0.1)

mtb = get_value_match(ab)
mti = get_value_match(ai)
mtf = get_value_match(af)

itb = get_value_if(ab)
iti = get_value_if(ai)
itf = get_value_if(af)
```

**VS Code extension or command-line**
VSCode 1.87.1 with extension, pyright v1.1.382. Pylance and CLI on file give same results.

**Comment by erictraut** [comment]
Pyright is working as intended here, so I don't consider this a bug.

The `isinstance` and `match` cases that you've provided are not equivalent, at least from the perspective of a static type checker.

You'll notice that in the `isinstance` case, the type of the `attribute` variable is not narrowed. This is also true in the `match` case where `attribute` is the subject expression for the `match` statement. The type of the subject expression is likewise not narrowed in this case.

Exhaustive match detection is based on type narrowing for the subject expression. Since the subject expression isn't narrowed in this case, exhaustion isn't detected, and an error is therefore reported.

In the case of your `isinstance` code sample, you're narrowing the expression `attribute.value`, and exhaustion checks are based on *its* narrowed type. 

The equivalent match statement in this case is the following, which type checks without errors.

```python
def get_value_match(attribute: Attribute[T]) -> T:
    match attribute.value:
        case bool() as vb:
            reveal_type(vb)
            return vb
        case int() as vi:
            reveal_type(vi)
            return vi
        case float() as vf:
            reveal_type(vf)
            return vf
```

I'll note that your use of a value-constrained type variable in this case is suspect because the three value constraints that you've provided entirely overlap each other — `bool` is a subtype of `int` which is a subtype of `float`. The typing spec doesn't say what a type checker should do in such a situation, but defining value-constrained type variables in this manner can lead to unsoundness. One could argue that it should be disallowed. You may want to consider replacing the value-constrained type variable with a normal type variable with an upper bound of `float` (as in `T = TypeVar("T", bound=float)`).

**Comment by GDYendell** [comment]
Thanks for the quick and thorough response! I think I understand the problem you outline above and I will look into refactoring this to not depend on the overlapping built-in types for type narrowing by making `DataType` generic over the `Bool`, `Int`, `String` classes instead.
